app/uk/gov/hmrc/cgtpropertydisposals/service/UpscanService.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.cgtpropertydisposals.service
+
+import java.time.LocalDateTime
+
+import cats.data.{EitherT, NonEmptyList}
+import cats.effect.{ContextShift, IO}
+import cats.implicits._
+import com.google.inject.{ImplementedBy, Inject, Singleton}
+import configs.Configs
+import configs.syntax._
+import play.api.Configuration
+import uk.gov.hmrc.cgtpropertydisposals.connectors.UpscanConnector
+import uk.gov.hmrc.cgtpropertydisposals.models.Error
+import uk.gov.hmrc.cgtpropertydisposals.models.dms.FileAttachment
+import uk.gov.hmrc.cgtpropertydisposals.models.ids.CgtReference
+import uk.gov.hmrc.cgtpropertydisposals.models.upscan.UpscanFileDescriptor.UpscanFileDescriptorStatus.UPLOADED
+import uk.gov.hmrc.cgtpropertydisposals.models.upscan.{FileDescriptorId, UpscanCallBack, UpscanFileDescriptor, UpscanSnapshot}
+import uk.gov.hmrc.cgtpropertydisposals.repositories.upscan.{UpscanCallBackRepository, UpscanFileDescriptorRepository}
+import uk.gov.hmrc.cgtpropertydisposals.util.Logging
+
+import scala.concurrent.{ExecutionContext, Future}
+@ImplementedBy(classOf[UpscanServiceImpl])
+trait UpscanService {
+
+  def getUpscanSnapshot(cgtReference: CgtReference): EitherT[Future, Error, UpscanSnapshot]
+
+  def storeFileDescriptorData(fd: UpscanFileDescriptor): EitherT[Future, Error, Unit]
+
+  def storeCallBackData(cb: UpscanCallBack): EitherT[Future, Error, Unit]
+
+  def getFileDescriptor(
+    cgtReference: CgtReference,
+    fileDescriptorId: FileDescriptorId
+  ): EitherT[Future, Error, Option[UpscanFileDescriptor]]
+
+  def getUpscanFileDescriptor(fileDescriptorId: FileDescriptorId): EitherT[Future, Error, Option[UpscanFileDescriptor]]
+
+  def updateUpscanFileDescriptorStatus(upscanFileDescriptor: UpscanFileDescriptor): EitherT[Future, Error, Boolean]
+
+  def getAllUpscanCallBacks(cgtReference: CgtReference): EitherT[Future, Error, List[UpscanCallBack]]
+
+  def downloadFilesFromS3(
+    snapshot: UpscanSnapshot,
+    urls: List[UpscanCallBack]
+  ): EitherT[Future, Error, List[Either[Error, FileAttachment]]]
```

Comment:
this return type is a bit weird - you've got errors in the right type as well. Would this not make more sense to be `EitherT[Future, Error, List[FileAttachment]]` and to collect the errors into one big one?

Reply:
discussed this morning - not sure if it resolves to any other type than this - do you want to take a deeper look while I crack on with try to do integration testing?

Reply:
let's do that then, yeah. Let's get this thing tested